enterprise/server/remote_execution/dirtools/dirtools_test.go

```diff
@@ -10,18 +11,322 @@ import (
 	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/dirtools"
 	"github.com/buildbuddy-io/buildbuddy/enterprise/server/remote_execution/filecache"
 	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/byte_stream_server"
+	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/content_addressable_storage_server"
 	"github.com/buildbuddy-io/buildbuddy/server/remote_cache/digest"
 	"github.com/buildbuddy-io/buildbuddy/server/testutil/testenv"
 	"github.com/buildbuddy-io/buildbuddy/server/testutil/testfs"
 	"github.com/buildbuddy-io/buildbuddy/server/util/hash"
 	"github.com/buildbuddy-io/buildbuddy/server/util/prefix"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
 	repb "github.com/buildbuddy-io/buildbuddy/proto/remote_execution"
 	rspb "github.com/buildbuddy-io/buildbuddy/proto/resource"
 	bspb "google.golang.org/genproto/googleapis/bytestream"
 )
 
+func TestUploadTree(t *testing.T) {
+	for _, tc := range []struct {
+		name           string
+		cmd            *repb.Command
+		directoryPaths []string
+		fileContents   map[string]string
+		symlinkPaths   map[string]string
+
+		expectedResult *repb.ActionResult
+		expectedInfo   *dirtools.TransferInfo
+	}{
+		{
+			name:           "NoFiles",
+			cmd:            &repb.Command{},
+			directoryPaths: []string{},
+			fileContents:   map[string]string{},
+			symlinkPaths:   map[string]string{},
+			expectedResult: &repb.ActionResult{},
+			expectedInfo: &dirtools.TransferInfo{
+				FileCount:        0,
+				BytesTransferred: 0,
+			},
+		},
+		{
+			name: "SomeFile",
+			cmd: &repb.Command{
+				OutputFiles: []string{"fileA.txt"},
+			},
+			directoryPaths: []string{},
+			fileContents: map[string]string{
+				"fileA.txt": "a",
+			},
+			symlinkPaths: map[string]string{},
+			expectedResult: &repb.ActionResult{
+				OutputFiles: []*repb.OutputFile{
+					{
+						Path: "fileA.txt",
+						Digest: &repb.Digest{
+							SizeBytes: 1,
+							Hash:      "ca978112ca1bbdcafac231b39a23dc4da786eff8147c4e72b9807785afee48bb",
+						},
+					},
+				},
+			},
+			expectedInfo: &dirtools.TransferInfo{
+				FileCount:        1,
+				BytesTransferred: 1,
+			},
+		},
+		{
+			name: "FileAndDirectory",
+			cmd: &repb.Command{
+				OutputFiles:       []string{"a/fileA.txt"},
+				OutputDirectories: []string{"a"},
```

Comment:
I might be misunderstanding but I think an `OutputDirectory` won't include an `OutputFile` as its child, otherwise the child will get uploaded twice. I think a more representative example would be to have "b" as an output directory, then in the fileContents map below, create a few files fileB.txt and fileC.txt